src/execution/sql/table_vector_iterator.cpp

```diff
@@ -23,41 +31,112 @@ bool TableVectorIterator::Init() {
   projected_columns_ = pc_init.Initialize(buffer_);
   initialized_ = true;
 
-  // Begin iterating
-  iter_ = std::make_unique<storage::DataTable::SlotIterator>(table_->begin());
+  // Create the start and end iterator
+  iter_ = std::make_unique<storage::DataTable::SlotIterator>(table_->beginAt(start_block_idx_));
+  iter_end_ = std::make_unique<storage::DataTable::SlotIterator>(table_->endAt(end_block_idx_));
   return true;
 }
 
 bool TableVectorIterator::Advance() {
   if (!initialized_) return false;
   // First check if the iterator ended.
-  if (*iter_ == table_->end()) {
+  if (*iter_ == *iter_end_) {
     return false;
   }
   // Scan the table to set the projected column.
-  table_->Scan(exec_ctx_->GetTxn(), iter_.get(), projected_columns_);
+  table_->RangeScan(exec_ctx_->GetTxn(), iter_.get(), iter_end_.get(), projected_columns_);
   pci_.SetProjectedColumn(projected_columns_);
   return true;
 }
 
 void TableVectorIterator::Reset() {
   if (!initialized_) return;
-  iter_ = std::make_unique<storage::DataTable::SlotIterator>(table_->begin());
+  // set the iterator to the start position specified by tbb
+  iter_ = std::make_unique<storage::DataTable::SlotIterator>(table_->beginAt(start_block_idx_));
 }
 
-bool TableVectorIterator::ParallelScan(uint32_t db_oid, uint32_t table_oid, void *const query_state,
-                                       ThreadStateContainer *const thread_states, const ScanFn scan_fn,
-                                       const uint32_t min_grain_size) {
-  // TODO(Amadou): Implement Me!!
-  return false;
+namespace {
+
+class ScanTask {
+ public:
+  ScanTask(exec::ExecutionContext *exec_ctx, uint16_t table_id, uint32_t *col_oids, uint32_t num_oids,
+           void *const query_state, TableVectorIterator::ScanFn scanner)
+      : exec_ctx_(exec_ctx),
+        table_id_(table_id),
+        col_oids_(col_oids),
+        num_oids_(num_oids),
+        query_state_(query_state),
+        scanner_(scanner) {}
+
+  void operator()(const tbb::blocked_range<uint32_t> &block_range) const {
+    // Create the iterator over the specified block range
+    TableVectorIterator iter(exec_ctx_, table_id_, col_oids_, num_oids_, block_range.begin(), block_range.end());
+
+    // Initialize the table vector iterator
+    if (!iter.Init()) {
+      return;
+    }
+
+    // Call scanning function which should be passed at runtime
+    scanner_(query_state_, exec_ctx_, &iter);
+  }
+
+ private:
+  exec::ExecutionContext *exec_ctx_;
+  uint16_t table_id_;
+  uint32_t *col_oids_;
+  uint32_t num_oids_;
+  void *const query_state_;
+  TableVectorIterator::ScanFn scanner_;
+};
+}  // namespace
+
+bool TableVectorIterator::ParallelScan(uint32_t table_oid, uint32_t *col_oids, uint32_t num_oids,
+                                       void *query_state, const ScanFn scan_fn, exec::ExecutionContext *exec_ctx,
+                                       unsigned int num_threads) {
```

Comment:
Is there a reason why `num_threads` is a `unsigned int` but `num_oids` is `uint32_t`?